proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,266 @@
+updated at each CPI call edge. The contents of this region are the following:
+
+- For each instruction in transaction:
+  - Index in transaction of program account to be executed: `u64`
+  - Reference to a slice of instruction accounts `&[InstructionAccount]`, 
+    consisting of:
+    - Pointer to beginning of slice: `u64`
+    - Number of elements in slice: `u64`
+  - Instruction data `&[u8]`, which is composed of:
+    - Pointer to data: `u64`
+    - Length of data: `u64`
+
+Let `InstrucionAccount` contain the following fields:
+
+  - Index to transaction account: `u16`
+  - Flags bitfield: `u16` (bit 0 is signer, bit 1 is writable)
+
+The second region is readonly and starts at `0x700000000`. It must also be 
+updated at each CPI call edge. This region contains an array of all 
+`InstructionAccounts` for every instruction in the transction. In other words:
+
+- For instruction in transaction:
+  - For each account in instruction:
+    - `InstructionAccount`, consisting of:
+      - Index to transaction account: `u16`
+      - Flags bitfield: `u16` (bit 0 is signer, bit 1 is writable)
+
+### Accounts area
+
+For each unique (meaning deduplicated) instruction account the payload must
+be mapped in at `0x800000000` plus `0x100000000` times the index of the
+**transaction** account (not the index of the instruction account). Only if the
+instruction account has the writable flag set and is owned by the current
+program it is mapped in as a writable region. The writability of a region must
+be updated as programs through out the transaction modify the account metadata.
+
+The runtime must only map the payload for accounts that belong in the current
+executing instruction. The payload for accounts belonging to sibling instructions
+must NOT be mapped.
+
+### Instruction payload area
+
+For each instruction, the runtime must map its payload at address 
+`0x10800000000` plus `0x100000000` times the index of the instruction in the 
+trasaction. All instruction payload mappings are readonly.
+
+One extra writable mapping must be created after the last instruction payload 
+area to be the CPI scratch pad, i.e. at address `0x10800000000` plus 
+`0x100000000` times the number of instructions in the transaction. Its purpose 
+is for programs to write CPI instruction data directly to it and avoid copies.
+
+### VM initialization
+
+During the initilization of the virtual machine, the runtime must load the 
+value `0x400000000` in register `R1` and value `0x600000000` in register `R2`. 
+These values represent addresses for programs to easily find the areas to read 
+information about the transaction and the instructions.
+
+### Changes to syscalls
+
+Changes to the account metadata must now be communicated with specific 
+syscalls, as detailed below:
+
+- `sol_resize_account`: Dst account, new length as `u64`
+- `sol_assign_owner`: Dst account, new owner as `&[u8; 32]`
+- `sol_transfer_lamports`: Dst account, src account, amount as `u64`
```

Comment:
Wonder if having "transfer" on the name here could create confusion with "transfer" in the system program – most likely not, but we could consider naming this as `sol_move_lamports`.

Reply:
It does the same thing as "Transfer" of the system program. But there is no risk of accidentally using the wrong one, because this one only works for accounts owned by the current program and the system program one only works for accounts owned by the system program.

Reply:
I was just thinking in a different name to emphasize that one is "lower" level than the other – e.g., the system program could implement its "Transfer" instruction using the `sol_move_lamports` syscall. The "move" one (probably?) does not require a signer check, since it can only move lamports out of accounts owned by the current program. But this is totally cosmetic.